XyzWalker/Assets/Src/Scripts/CodeExecutor.cs

```diff
@@ -47,55 +47,39 @@ private class ConsoleWriter : TextWriter {
     }
   }
 
+  public bool IsRunning => !_engine.IsStopped;
+
   private const string _defaultModuleName = "Program";
   private const int _minSleepInMilliSeconds = 10;
   private const float _singleStepWaitInSeconds = .1f;
 
   private readonly GameManager _gameManager;
   private readonly ConsoleWriter _consoleWriter;
-  private readonly object _threadLock = new object();
-
-  // If the executor thread is running.
-  public bool IsRunning => !(_thread is null);
+  private readonly Engine _engine = new Engine(SeedXLanguage.SeedPython, RunMode.Script);
 
-  // A switch for the caller to stop the code execution when the thread is running. It's safe for
-  // another thread to flip this bool flag directly.
+  // A switch for the caller to stop the code execution when the coroutine is running.
   public bool Stopping = false;
 
-  private Thread _thread;
-  private string _source;
-
   public CodeExecutor(GameManager gameManager) {
     _gameManager = gameManager;
     _consoleWriter = new ConsoleWriter(gameManager);
+    _engine.RedirectStdout(_consoleWriter);
+    _engine.Register(this);
   }
 
   // Runs a SeedLang script. Returns false if the executor is already running.
   public bool Run(string source) {
-    bool ret = true;
-    lock (_threadLock) {
-      if (_thread is null) {
-        _source = source;
-        Stopping = false;
-        _thread = new Thread(ThreadEntry);
-        _thread.Start();
-      } else {
-        ret = false;
-      }
+    if (IsRunning) {
+      return false;
     }
-    return ret;
+    _gameManager.StartCoroutine(RunProgram(source));
+    return true;
   }
 
   public void On(Event.SingleStep e, IVM vm) {
-    // Checks the stopping flag in the SingleStep callback.
-    if (Stopping) {
-      vm.Stop();
-      Stopping = false;
-    } else {
-      // Highlights the current line.
-      _gameManager.QueueHighlightCodeLineAndWait(e.Range.Start.Line, _singleStepWaitInSeconds);
-      WaitForActionQueueComplete();
-    }
+    // Highlights the current line.
+    _gameManager.QueueHighlightCodeLineAndWait(e.Range.Start.Line, _singleStepWaitInSeconds);
+    vm.Pause();
```

Comment:
This coroutine solution leverages the SingleStep callback to call vm.Pause() to release the execution ownership.

This solution is perfect for the example projects such as XyzWalker and SortingBot. But it has a problem for future use cases that use the SeedLang engine as a pure animation controller.

Consider this case: the Unity app starts SeedLang to calculate the position and the rotation of a number of objects, storing the calculation results to shared variables. The main animation loop reads the shared variables and draws animations. It's required that the SeedLang engine runs in as high performance as possible. This leads to two issues:

Issue 1) Listening to SingleStep events will generate SingleStep instructions after compilation. Those SingleStep instructions and the corresponding event handler invocation operations will slow down the running of the engine.

Issue 2) When the SeedLang VM is running in Unity's main thread, it's ideal that it takes a reasonable time during the execution of one frame (it takes only a part of 1/60 sec, for example). How long should it take in this circumstance? How many instructions should it execute before it calls vm.Pause() to return the execution ownership? This is SeedLang engine's own duty. If using the SingleStep event to call vm.Pause(), the client code must be responsible for maintaining a counter and determining how often (i.e., every N single steps) to call vm.Pause(). It doesn't sound good to let the client code to set the very critical option (the frequency to call vm.Pause()) for the SeedLang engine, right?

Reply:
I have compared the tradeoff of different solutions, some of them may need more efforts to think about and implement.

1. SeedLang provides an interface to run a given numbers of bytecode or a certain time (like 10ms provided by the client). The implement code looks like:
```csharp
while (true) {
    RunABytecode();
    count++;     // or counts execution time
    if (count > expectedBytecodeNumbers) {
        count = 0;
        return;
    }
}
```
The execution performance will be several times slower than before.

2. Introduce a partial running mode based on solution 1. The code will be:
```csharp
while (true) {
    RunABytecode();
    if (IsInPartialRunningMode()) {
        count++;     // or counts execution time
        if (count > expectedBytecodeNumbers) {
            count = 0;
            return;
        }
    }
}
```
The normal mode will still be a little slower by one partial running mode check.

3. SeedLang can add a SingleStep observer by itself, and check the source line number or execution time in the SingleStep observer, pause and exit the execution if condition is met.

Maybe in the future, we need do tradeoff of these options and implement a mechanism in SeedLang to measure and limit the execution time of each running call. Current solution is just a workaround of option 3.

Reply:
Option 4 (not a replacement of the above options. Instead, this can be an additional solution): Support C#/Unity's `yield` syntax in SeedPython and let the script author decide when and where to return the execution owner to the main loop. For example, a computing-intensive SeedPython script can implement the following logic:

```python
for x in range(10000):
    sum += complex_computing_on(x)
    yield None  # This returns the execution logic to Unity's main loop and waits for the next frame.

print(sum)
```

Reply:
Okay to go with the current simple solution and we can improve SeedLang's coroutine ability later.